File: pfcomponent/aspx/protect/Pfsecure.cs
/**
 * This file is covered by the GNU GPLv3 license <http://www.gnu.org/licenses/gpl.html>
 * To purchase a commercial license at http://w3plan.net
 * 
 * Copyright 2015-2016 W3plan Technologies, http://w3plan.net
 *
 */

using System.IO;
using System;
using System.Net;

class Pfsecure {
	/**
	 * Encrypt data and return result
	 * 
	 * @param {string} str     Plain text
	 * @param {string} pword   password for encryption
	 * @return {string}        Cypher text
	 */
	public string pfEncryption(string str, string pword) {
		try {
			string val = WebUtility.UrlEncode(str);				
			val = Xor.Pfxor(val, pword);
			byte[] sval = System.Text.Encoding.UTF8.GetBytes(val);		
			// encode special character "+", "/" and "=" in base64 result
			return WebUtility.UrlEncode(System.Convert.ToBase64String(sval));		
		} catch (Exception e) {
			Console.WriteLine("Encryption failed, {0} ", e);
			return "";
		}
	}

	/**
	 * Decrypted data and return result
	 * 
	 * @param {string}  str     XOR encrypted data
	 * @param {string}  pword   password for decryption
	 * @return {string} plain text      
	 */
	public string pfDecryption(string str, string pword) {
		try {
			// encode special character "+", "/" and "=" in str
			string val = WebUtility.UrlDecode(str);
			byte[] sVal = System.Convert.FromBase64String(val);
			string bVal = Xor.Pfxor(System.Text.Encoding.UTF8.GetString(sVal), pword);
			return WebUtility.UrlDecode(bVal);		
		} catch (Exception e) {
			Console.WriteLine("Decryption failed, {0} ", e);
			return "";
		}
	}
	
	/**
	 * Entry
	 */
	static void Main() {     
		
    }
}

class Xor {
    /**
	 * Encryption and decryption ASCII string with XOR cipher and an optional password for C#
	 * 
	 * @public
	 * @param {string} instr   	 ASCII string to be encrypted/decrypted
	 * @param {string=} pword  	 Optional XOR password with 8 characters at least, 
	 *							 default length of password is 256
	 * @return {string}   		 Encrypted/decrypted string, return an empty string if instr is empty 
	 */
   public static string Pfxor (string instr, string pword) {
		if (instr.Trim().Length < 1) {
			return "";
		}
		
		if (pword.Length < 8) {
			pword = "8q2EXsYBCDZcK3hajFFnyzdGRTbnrYbjYJzkbMFfRJFdvhPfBmpNVw2YkBZtM9kLW6MRAst7Vb3yh8KZwq2dTNuVdq8acHYeavBaPz3MPsBGpAP3zaCDvZUTvNGaWvpNwqwnQ9D8nZ8T4K9D8HRyQ2XTapaAeDSUfanvkCkRFzh4vSs3C9qBWxTwx9PUTTrAaL5PfgvQRWaCtCAZng3P8S9aEYEST79w2Ryu5Vs4etvKz4xdM8K7uCn2yFZ5C2MJ";
		}
		
		int icnt = instr.Length;
		int index = 0;
		int i = 0;
		string retval = "";	
		for (i = 0; i < icnt; i++) {
			index = i % pword.Length;
			retval += instr[i] ^ pword[index];
		}
		
		return retval;
	}
}

File: pfcomponent/ruby/protect/pfsecure.rb
##
 # This file is covered by the GNU GPLv3 license <http://www.gnu.org/licenses/gpl.html>
 # To purchase a commercial license at http://w3plan.net
 # 
 # Copyright 2015-2016 W3plan Technologies, http://w3plan.net
 #
 ##

require "uri"
require "base64"
require "zlib"
 
class Pfsecure
	##
	 # Encrypt data and return result
	 # 
	 # @param {string} $str    Plain text
	 # @param {string} $pword  password for encryption
	 # @return {string}        Cypher text
	 ##
	def pfEncryption(str, pword)
		begin
			val = URI.encode(str) 		
			xVal = pfXor(val, pword)
			bVal = Base64.encode64(xVal)
			return URI.encode(bVal)
		rescue
			puts "Decryption failed"
		end
	end
	
	##
	 # Decrypted data and return result
	 # 
	 # @param {string}  $str    XOR encrypted data
	 # @param {string}  $pword  password for decryption
	 # @return {string} plain text      
	 ##
	def pfDecryption(str, pword)
		begin
			val = URI.decode(str)
			xVal = Base64.decode64(val)
			bVal = pfXor(xVal, pword)
			return URI.decode(bVal)
		rescue
		  puts "Decryption failed"
		end
	end

	 ##
	 # Encryption and decryption ASCII string with XOR cipher and an optional password for Ruby
	 # 
	 # @public
	 # @param {string} instr   	 ASCII string to be encrypted/decrypted
	 # @param {string=} pword  	 Optional XOR password with 8 characters at least, 
	 #							 default length of password is 256
	 # @return {string}			 Encrypted/decrypted string, return an empty string if instr is empty
	 #
	 ##
	def pfXor(instr, pword)
		if (instr.strip.length < 1)
			return ""
		end	
		if (pword.length < 8) 
			pword = "8q2EXsYBCDZcK3hajFFnyzdGRTbnrYbjYJzkbMFfRJFdvhPfBmpNVw2YkBZtM9kLW6MRAst7Vb3yh8KZwq2dTNuVdq8acHYeavBaPz3MPsBGpAP3zaCDvZUTvNGaWvpNwqwnQ9D8nZ8T4K9D8HRyQ2XTapaAeDSUfanvkCkRFzh4vSs3C9qBWxTwx9PUTTrAaL5PfgvQRWaCtCAZng3P8S9aEYEST79w2Ryu5Vs4etvKz4xdM8K7uCn2yFZ5C2MJ"
		end	
		keys = pword.split(//)	
		output = ""
		i = 0
		while c = instr.getc
			pass_char = keys[i]		
			xor = c.chr[0] ^ pass_char[0]		
			output += xor.chr		
			i+=1		
			if i == (keys.size - 1)		
				i = 0		
			end		
		end	
		return output	
	end
	
	##
	 # Check data weather matching gave CRC32 checksum
	 # 
	 # @public 
	 # @param {string} val   Original data
	 # @param {string} csm   decimal string of CRC32 checksum 
	 # @return {boolean}     True if data matches checksum otherwise false
	 ## 
	def pfCheckCrc(val, csm)
		if crc32CheckSum(val) == csm:
			return true
		else
			return false
		end
	end
	
	##
	 # Calculates a string's CRC32 checksum to check data integrity to see
	 # weather the data has been modified or changed 
	 #
	 # @public
	 # @param {string} str   A string in Latin1 (ISO8859-1) character set
	 # @return {string}	     A decimal string
	 ##
	def crc32CheckSum(str)
		return Zlib::crc32(str)
	end
end
